### PR TITLE
Version 1 - 116 improve output data formatting

### LINKED ISSUE #65: converted steps_string that is saved as metadata may lose information on functions objects passed to step classes
**Describe the bug**
When passing functions to steps like this:
`step = {'correct background': pyopia.background.CorrectBackgroundAccurate(pyopia.background.shift_bgstack_accurate)}`

then the converted step_string that is saved as human-readable metadata to the hdf5 file does not correctly capture the `pyopia.background.shift_bgstack_accurate` object

**Additional context**
It might be better to not be saving 'human-readable' metadata to the hdf5 file (but I can't get the HDF5 writer to accept the raw steps dict) and instead use something that can be read by machine and converted to human readable when needed so that we don't loose these type of details.

### LINKED ISSUE #96: pixel size should be included in output files
... as this will vary between instruments

### LINKED ISSUE #133: for holograms - Initial step requires a filename - maybe better to take the general raw files folder and read the first file from that
for holograms - Initial step requires a filename - maybe better to take the general raw files folder and read the first file from that?

_Originally posted by @animmosmith in https://github.com/SINTEF/pyopia/issues/118#issuecomment-1698217784_

### LINKED ISSUE #134: Check for consistency across documentation and examples

### LINKED ISSUE #139: make cli command that updates an existing config.toml file
instead of using the instrument.holo.generate_config, this function would have the same sort of input arguments, but use a toml file to get the other settings, rather than the source code.

### PR DESCRIPTION
**Pull request by emlynjdavies**
# This is a PR to progress the main branch to Version 1.

This is a summary of where i ended up with the interpretation of the steps dictionary from a [toml](https://toml.io/en/) [configuration file](https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/config.toml), and saving it to nc (instead of h5), and including a steps string that is possible to re-evaluate to re-run the analysis:
[notebooks/single-image-stats.ipynb](https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/single-image-stats.ipynb)

### Completed issues contained in this PR
- fixes #133
- fixes #96 
- fixes #65
- fixes #139
- fixes #134

----
## Holo test assertion change:
Number of particles expected in this test is now 56 instead of 42. This change is likely due to different processing settings used in the test. See [pyopia/tests/test_pipeline.py](https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/pyopia/tests/test_pipeline.py#L118)

----
Known issue with Mac M2 dependencies. For now, please use the https://github.com/SINTEF/pyopia/tree/mac-m branch if you don't have an intel mac

### CODE REVIEW

File: pyopia/io.py [hunk 1]
```diff
@@ -79,3 +98,40 @@ def __call__(self,
...
         return data
+
+
+def load_toml(toml_file):
+    with open(toml_file, 'r') as f:
+        settings = toml.load(f)
+    return settings
+
+
+def build_repr(settings, step_name):
```
Comment (emlynjdavies) [inline pyopia/io.py:109]: This should probably be in {mod}`pyopia.pipeline`

File: pyopia/io.py [hunk 2]
```diff
@@ -79,3 +98,40 @@ def __call__(self,
...
+
+    if len(arg_names) > 0:
+        repr_string = repr_string[:-2]
+    repr_string += ')'
+
+    output = f'\'{step_name}\': {repr_string}'
+    return output
+
+
+def build_step_string(settings):
```
Comment (emlynjdavies) [inline pyopia/io.py:128]: This should probably be in `pyopia.pipeline`

**Comment by emlynjdavies** [comment]
This method will have problems when it comes to handling steps that contain inputs that are not available in a config file e.g.

`'create background': pyopia.background.CreateBackground(bgfiles,
                                                                 pyopia.instrument.silcam.load_image)`

@animmosmith  would it be acceptable to insist that pipeline classes only take inputs of strings or floats that don't require computation (i.e. can be put in a config file)? this would mean a bit of refactoring?

Or are there other alternatives?

**Comment by animmosmith** [comment]
> This method will have problems when it comes to handling steps that contain inputs that are not available in a config file e.g.
> 
> `'create background': pyopia.background.CreateBackground(bgfiles, pyopia.instrument.silcam.load_image)`
> 
> @animmosmith would it be acceptable to insist that pipeline classes only take inputs of strings or floats that don't require computation (i.e. can be put in a config file)? this would mean a bit of refactoring?
> 
> Or are there other alternatives?

I think reworking things to make it work with a config file is worth it. I guess with the upgrade to 3.10 we can make use of match-case constructs to enable easy switching between difference processing functions?

**Comment by animmosmith** [comment]
> Getting back to this now, after the summer...
> 
> In the case of the background creation, we currently do:
> 
> `bgfiles = files[:average_window] # provide a list of background files used to create the static background`
> 
> I am thinking that we will need to change this to a folder name, which the background creator uses to find bgfiles. so it would then have the folder path and average_window as inputs in the config.
> 
> Does that sound OK @animmosmith ?

Yes, I think that would be easy to explain to end users if using the static background option - an additional option to the background creator tool could be either how many of the files in the folder to use (equally distributed) or what the stride length should be if sub-sampling the files in the (perhaps raw data) folder for speed.

**Comment by emlynjdavies** [comment]
@animmosmith How is this looking from your perspective now?

Specifically, the structure in the example [configuration file](https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/config.toml) (currently setup for silcam), and the new output format & metadata:
https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/single-image-stats.ipynb

Are we ready to see how this can work for holo analysis?

**Comment by animmosmith** [comment]
> @animmosmith How is this looking from your perspective now?
> 
> Specifically, the structure in the example [configuration file](https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/config.toml) (currently setup for silcam), and the new output format & metadata: https://github.com/SINTEF/pyopia/blob/116-improve-output-data-formatting/notebooks/single-image-stats.ipynb
> 
> Are we ready to see how this can work for holo analysis?

Struggling to get this to run. When using single-image-stats.ipynb, I get a "ModuleNotFoundError: No module named 'silcam'"  at the CreateBackground step....

**Comment by animmosmith** [comment]
OK - worked through the silicam example - I'll generate an equivalent holo settings and test

**Comment by animmosmith** [comment]
I think there is some further optimisation of the pipeline that would be helpful:

for holograms - Initial step requires a filename - maybe better to take the general raw files folder and read the first file from that?

for both instruments - create background step takes an instrument_module - and then the load step also uses essentially the same information - better to remove this duplication?

**Comment by animmosmith** [comment]
tests need updating to fit with new methods...

**Comment by emlynjdavies** [comment]
> I think there is some further optimisation of the pipeline that would be helpful:
> 
> for holograms - Initial step requires a filename - maybe better to take the general raw files folder and read the first file from that?
> 
> for both instruments - create background step takes an instrument_module - and then the load step also uses essentially the same information - better to remove this duplication?

Yep - agree.

**Comment by emlynjdavies** [comment]
> tests need updating to fit with new methods...

Yes - I can update testing once we are happy thing are working how we want

**Comment by emlynjdavies** [comment]
It seems that things worked OK for holo - but that we should do the updates you suggest that reduce duplication in the config by using the 'general' settings.

Is the threshold in the focus step different from the segmentation threshold?

**Review by animmosmith — APPROVED** [review]

**Comment by animmosmith** [comment]
> Is the threshold in the focus step different from the segmentation threshold?

Yes - they can be set to different values that may depend on which stack summary method is used and also on the reconstruction parameters that lighten or darken the background

**Comment by animmosmith** [comment]
looking at updating the pipeline-holo.ipyb now...

**Comment by animmosmith** [comment]
@emlynjdavies - can you please check the pipeline-holo to see if the explanation of the new pipeline config is clear?

**Comment by emlynjdavies** [comment]
@animmosmith  in the holo pipeline notebook, increase_depth_of_field is defined and then overwritten in the config dict. do you want this to be True or False?

**Comment by emlynjdavies** [comment]
Have discovered and installation inconsistency - need to test if this is only on Mac M1 chips....

**Comment by emlynjdavies** [comment]
> Have discovered and installation inconsistency - need to test if this is only on Mac M1 chips....

Have tested on intel mac, and poetry installation works and pytest passes. I think we might need to come back to resolving dependencies for Silicon chips later, as a quick-fix isn't immediately obvious.

**Comment by emlynjdavies** [comment]
STATS are not being appended to the NetCDF files, so we're only getting the last file's data from the pipeline.

**Review by animmosmith — APPROVED** [review]

**Pull request merged**